File: src/main/java/io/github/tanguygab/conditionalactions/actions/ActionManager.java
package io.github.tanguygab.conditionalactions.actions;

import io.github.tanguygab.conditionalactions.ConditionalActions;
import io.github.tanguygab.conditionalactions.actions.types.Action;
import io.github.tanguygab.conditionalactions.actions.types.commands.GroupAction;
import io.github.tanguygab.conditionalactions.actions.types.bungee.ServerAction;
import io.github.tanguygab.conditionalactions.actions.types.commands.*;
import io.github.tanguygab.conditionalactions.actions.types.data.*;
import io.github.tanguygab.conditionalactions.actions.types.items.*;
import io.github.tanguygab.conditionalactions.actions.types.messages.*;
import io.github.tanguygab.conditionalactions.actions.types.other.DelayAction;
import io.github.tanguygab.conditionalactions.actions.types.storage.*;
import io.github.tanguygab.conditionalactions.events.ActionsRegisterEvent;
import lombok.Getter;
import net.kyori.adventure.text.Component;
import org.bukkit.OfflinePlayer;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.entity.Player;

import java.io.File;
import java.util.*;
import java.util.regex.Matcher;

public class ActionManager {

    @Getter private final String argumentSeparator;
    private final ConditionalActions plugin;
    private final List<Action> actions = new ArrayList<>();
    private final Map<String,ActionGroup> actionGroups = new HashMap<>();

    public ActionManager(ConditionalActions plugin, String argumentSeparator) {
        this.argumentSeparator = argumentSeparator;
        this.plugin = plugin;
        register(
                new PlayerAction(),
                new ConsoleAction(),
                new GroupAction(),

                new ServerAction(),

                new RemoveDataAction(),
                new SetDataAction(),

                new EnchantItemAction("add"),
                new EnchantItemAction("set"),
                new EnchantItemAction("take"),
                new GiveItemAction(),
                new GiveItemStorageAction(),
                new RepairItemAction(),
                new TakeItemAction(),
                new TakeItemStorageAction(),

                new ActionBarAction(),
                new BroadcastAction(),
                new BroadcastActionBarAction(),
                new BroadcastTitleAction(),
                new ChatAction(),
                new MessageAction(),
                new TitleAction(),

                new DelayAction()
        );
        if (plugin.getServer().getPluginManager().isPluginEnabled("LuckPerms"))
            register(new PermissionAction());

        try {
            Player.class.getMethod("sendMessage", Component.class);
            register(new MiniMessageAction());
        } catch (NoSuchMethodException ignored) {}

        plugin.sync(null, this::load);
    }

    public void load() {
        ActionsRegisterEvent event = new ActionsRegisterEvent();
        plugin.getServer().getPluginManager().callEvent(event);
        registerAll(event.getActions());

        File file = new File(plugin.getDataFolder(),"actiongroups.yml");
        if (!file.exists()) plugin.saveResource("actiongroups.yml",false);
        YamlConfiguration groups = YamlConfiguration.loadConfiguration(file);
        groups.getValues(false).forEach((name,config)->{
            if (config instanceof List<?> list) actionGroups.put(name,new ActionGroup(this,list));
        });

        plugin.getCustomCommandManager().getCommands().values().forEach(command->command.loadActions(this));
    }

    public boolean findAndExecute(OfflinePlayer player, String action, boolean group) {
        CAExecutable executable;

        if (!group) {
            Action ac = find(action);
            executable = ac == null ? null : new ActionData(ac,action);
        } else executable = actionGroups.get(action);

        if (executable == null) return false;

        if (plugin.getServer().isPrimaryThread())
            ConditionalActions.getInstance().async(()->executable.execute(player));
        else executable.execute(player);

        return true;
    }

    public Action find(String action) {
        for (Action ac : actions) {
            Matcher matcher = ac.getPattern().matcher(action);
            if (matcher.find()) return ac;
        }
        return null;
    }

    public void register(Action... action) {
        actions.addAll(List.of(action));
    }
    public void registerAll(List<Action> action) {
        actions.addAll(action);
    }

    public List<String> getSuggestions() {
        return actions.stream().map(Action::getSuggestion).toList();
    }

    public Set<String> getGroups() {
        return actionGroups.keySet();
    }
}

File: src/main/kotlin/io/github/tanguygab/conditionalactions/actions/ActionManager.kt
package io.github.tanguygab.conditionalactions.actions

import io.github.tanguygab.conditionalactions.ConditionalActions
import io.github.tanguygab.conditionalactions.Utils
import io.github.tanguygab.conditionalactions.actions.types.bungee.BungeeConsoleAction
import io.github.tanguygab.conditionalactions.actions.types.bungee.BungeePlayerAction
import io.github.tanguygab.conditionalactions.actions.types.items.EnchantItemAction
import io.github.tanguygab.conditionalactions.actions.types.items.GiveItemAction
import io.github.tanguygab.conditionalactions.actions.types.items.RepairItemAction
import io.github.tanguygab.conditionalactions.actions.types.items.TakeItemAction
import io.github.tanguygab.conditionalactions.actions.types.storage.GiveItemStorageAction
import io.github.tanguygab.conditionalactions.actions.types.storage.TakeItemStorageAction
import io.github.tanguygab.conditionalactions.actions.types.bungee.ServerAction
import io.github.tanguygab.conditionalactions.actions.types.commands.*
import io.github.tanguygab.conditionalactions.actions.types.data.*
import io.github.tanguygab.conditionalactions.actions.types.messages.*
import io.github.tanguygab.conditionalactions.actions.types.other.*
import io.github.tanguygab.conditionalactions.events.ActionsRegisterEvent

import org.bukkit.OfflinePlayer
import org.bukkit.configuration.file.YamlConfiguration

class ActionManager(private val plugin: ConditionalActions, val argumentSeparator: String) {
    private val actions = mutableListOf<Action>()
    private val actionGroups = mutableMapOf<String, ActionGroup>()

    init {
        register(
            PlayerAction(),
            ConsoleAction(),
            GroupAction(),

            BungeePlayerAction(),
            BungeeConsoleAction(),
            ServerAction(),

            RemoveDataAction(),
            SetDataAction(),

            EnchantItemAction("add"),
            EnchantItemAction("set"),
            EnchantItemAction("take"),
            GiveItemAction(),
            GiveItemStorageAction(),
            RepairItemAction(),
            TakeItemAction(),
            TakeItemStorageAction(),

            ActionBarAction(),
            BroadcastAction(),
            BroadcastActionBarAction(),
            BroadcastTitleAction(),
            ChatAction(),
            MessageAction(),
            TitleAction(),

            PlaceholderAction(),
            LogAction(),
            DelayAction()
        )
        if (plugin.server.pluginManager.isPluginEnabled("LuckPerms")) register(
            PermissionAction(),
            GiveTakePermissionAction("give"),
            GiveTakePermissionAction("take")
        )

        plugin.sync(null) { load() }
    }

    fun load() {
        val event = ActionsRegisterEvent()
        plugin.server.pluginManager.callEvent(event)
        registerAll(event.actions)

        Utils.updateFiles(plugin, "actiongroups.yml", "actiongroups/default-groups.yml")
        Utils.loadFiles(plugin.dataFolder.resolve("actiongroups"), "") { file, _ ->
            YamlConfiguration.loadConfiguration(file).getValues(false).forEach { (name, obj) ->
                if (obj is List<*>) actionGroups[name] = ActionGroup(this, obj)
            }
        }

        plugin.customCommandManager.commands.values.forEach { it.loadActions(this) }
    }

    fun findAndExecute(player: OfflinePlayer?, action: String, group: Boolean): Boolean {
        val executable: CAExecutable?

        if (!group) {
            val ac = find(action)
            executable = if (ac == null) null else ActionData(ac, action)
        } else executable = actionGroups[action]

        if (executable == null) return false

        if (plugin.server.isPrimaryThread) ConditionalActions.INSTANCE.async { executable.execute(player) }
        else executable.execute(player)

        return true
    }

    fun find(action: String) = actions.find { it.pattern.containsMatchIn(action) }

    fun register(vararg action: Action) = actions.addAll(action)

    fun registerAll(action: List<Action>) = actions.addAll(action)

    fun getSuggestions() = actions.map { it.getSuggestion() }

    fun getGroups() = actionGroups.keys
}
